kitsune/wiki/drafts.py

```diff
@@ -0,0 +1,186 @@
+        except DraftRevision.DoesNotExist:
+            raise ValueError(f"Draft {draft_id} not found for user {user.username}")
+
+        return {
+            "title": draft.title,
+            "slug": draft.slug,
+            "content": draft.content,
+            "summary": draft.summary,
+            "keywords": draft.keywords,
+            "based_on": draft.based_on.id if draft.based_on else None,
+        }
+
+    def discard_draft(self, draft_id: int, user) -> bool:
+        """Delete a draft revision.
+
+        Args:
+            draft_id: ID of the draft to delete
+            user: The user requesting deletion
+        Returns:
+            bool: True if draft was successfully deleted
+        """
+        try:
+            draft = DraftRevision.objects.get(id=draft_id, creator=user)
+        except DraftRevision.DoesNotExist:
+            return False
+        else:
+            draft.delete()
+            return True
+
+    def create_revision(self, form_data: dict[str, Any], creator, document, based_on_id=None, base_rev=None, send_notifications=False) -> Revision:
+        """Create a new revision from form data with validation and permissions.
+
+        Args:
+            form_data: Dictionary containing revision data (content, summary, keywords, etc.)
+            creator: The user creating the revision
+            document: The document this revision belongs to
+            based_on_id: Optional ID of the revision this is based on
+            base_rev: Optional base revision object for keyword permission checking
+            send_notifications: Whether to send content creation notifications
+        Returns:
+            Revision: The created revision
+        """
+        revision = Revision(**form_data)
+        revision.document = document
+        revision.creator = creator
+
+        if based_on_id:
+            revision.based_on_id = based_on_id
+
+        # Handle keyword permissions - if user can't edit keywords, preserve old value
+        if base_rev and not document.allows(creator, "edit_keywords"):
+            revision.keywords = base_rev.keywords
+
+        revision.save()
+        if send_notifications:
+            self.fire_notifications(revision, [NotificationType.CONTENT_CREATION])
+        return revision
+
+    def publish_draft(self, draft_id: int, user, form_data: dict[str, Any], based_on_id=None, base_rev=None) -> Revision:
+        """Convert a draft to a published revision and delete the draft.
+
+        Args:
+            draft_id: ID of the draft to publish
+            user: The user publishing the draft
+            form_data: Form data for revision creation
+            based_on_id: Optional ID of the revision this is based on
+            base_rev: Optional base revision object
+        Returns:
+            Revision: The created revision
+        """
+        try:
+            draft = DraftRevision.objects.get(id=draft_id, creator=user)
+            document = draft.document
```

Comment:
To get rid of an extra query, maybe something like the following instead?

```python
draft = DraftRevision.objects.select_related("document").get(id=draft_id, creator=user)
```

and then use `draft.document` below?